x-pack/plugin/async-search/src/main/java/org/elasticsearch/xpack/search/TransportGetAsyncMyStatusAction.java

```diff
@@ -0,0 +1,136 @@
+            EsExecutors.DIRECT_EXECUTOR_SERVICE
+        );
+        this.transportService = transportService;
+        this.clusterService = clusterService;
+        this.store = new AsyncTaskIndexService<>(
+            XPackPlugin.ASYNC_RESULTS_INDEX,
+            clusterService,
+            threadPool.getThreadContext(),
+            client,
+            ASYNC_SEARCH_ORIGIN,
+            AsyncSearchResponse::new,
+            registry,
+            bigArrays
+        );
+    }
+
+    @Override
+    protected void doExecute(Task task, GetAsyncStatusRequest request, ActionListener<AsyncStatusResponse> listener) {
+        AsyncExecutionId searchId = AsyncExecutionId.decode(request.getId());
+        DiscoveryNode node = clusterService.state().nodes().get(searchId.getTaskId().getNodeId());
+        DiscoveryNode localNode = clusterService.state().getNodes().getLocalNode();
+
+        if (node == null || Objects.equals(node, localNode)) {
+            if (request.getKeepAlive() != null && request.getKeepAlive().getMillis() > 0) {
+                long expirationTime = System.currentTimeMillis() + request.getKeepAlive().getMillis();
+                store.updateExpirationTime(searchId.getDocId(), expirationTime, ActionListener.wrap(p -> {
+                    AsyncSearchTask asyncSearchTask = store.getTaskAndCheckAuthentication(taskManager, searchId, AsyncSearchTask.class);
+                    if (asyncSearchTask != null) {
+                        asyncSearchTask.setExpirationTime(expirationTime);
+                    }
+                    store.retrieveStatus(
+                        request,
+                        taskManager,
+                        AsyncSearchTask.class,
+                        AsyncSearchTask::getStatusResponse,
+                        AsyncStatusResponse::getStatusFromStoredSearch,
+                        listener
+                    );
+                }, exc -> {
+                    RestStatus status = ExceptionsHelper.status(ExceptionsHelper.unwrapCause(exc));
+                    if (status != RestStatus.NOT_FOUND) {
+                        logger.error(() -> format("failed to update expiration time for async-search [%s]", searchId.getEncoded()), exc);
+                        listener.onFailure(exc);
+                    } else {
+                        // the async search document or its index is not found.
+                        // That can happen if an invalid/deleted search id is provided.
+                        listener.onFailure(new ResourceNotFoundException(searchId.getEncoded()));
+                    }
+                }));
+            } else {
+                // ensure that the user requesting status is the same one that created the search
+                // TODO: this works only while the task is running - once it completes, other users can access the status
+                try {
+                    store.getTaskAndCheckAuthentication(taskManager, searchId, AsyncSearchTask.class);
```

Comment:
This method is copied from `TransportGetAsyncStatusAction.doExecute`, except I added this additional security check for when the `keep_alive` flag is not present.

Based on my manual testing with two different users (both with `read` permissions), while the search is running, only the user that created the search can see the status via this code (including when the keep_alive flag is present since that code does the same permissions check).

However, when the task is finished and the status is retrieved from disk this check no longer works and the other user can see the status.

How do we enforce security after the task is done?

I tried tracking how the `GET _async_search` endpoint does it and got the following stack trace (when querying it via a user that did not create the task):

<details><summary><b>Stack trace</b> (toggle to view)</summary>

```
$ curlhj -X GET "https://localhost:9200/_async_search/$SID?error_trace=true" --cacert primary-http_ca.crt -u xx_user:$MYPASS

{"error":{"root_cause":[{"type":"resource_not_found_exception","reason":"FlplSzBxMkxwUzBTSWpOaXotYnlvVVEbZmhLUk8zWG9RZ2lJM2tKZGVtX3d4ZzozMTYy",
"stack_trace":"org.elasticsearch.ResourceNotFoundException: FlplSzBxMkxwUzBTSWpOaXotYnlvVVEbZmhLUk8zWG9RZ2lJM2tKZGVtX3d4ZzozMTYy
	at org.elasticsearch.xcore@8.14.0-SNAPSHOT/org.elasticsearch.xpack.core.async.AsyncTaskIndexService.parseResponseFromIndex(AsyncTaskIndexService.java:476)
	at org.elasticsearch.xcore@8.14.0-SNAPSHOT/org.elasticsearch.xpack.core.async.AsyncTaskIndexService.lambda$getResponseFromIndex$5(AsyncTaskIndexService.java:439)
	at org.elasticsearch.server@8.14.0-SNAPSHOT/org.elasticsearch.action.ActionListenerImplementations$DelegatingFailureActionListener.onResponse(ActionListenerImplementations.java:217)
	at org.elasticsearch.server@8.14.0-SNAPSHOT/org.elasticsearch.action.support.ContextPreservingActionListener.onResponse(ContextPreservingActionListener.java:32)
	at org.elasticsearch.server@8.14.0-SNAPSHOT/org.elasticsearch.tasks.TaskManager$1.onResponse(TaskManager.java:202)
	at org.elasticsearch.server@8.14.0-SNAPSHOT/org.elasticsearch.tasks.TaskManager$1.onResponse(TaskManager.java:196)
	at org.elasticsearch.server@8.14.0-SNAPSHOT/org.elasticsearch.action.ActionListenerImplementations$RunBeforeActionListener.onResponse(ActionListenerImplementations.java:307)
	at org.elasticsearch.server@8.14.0-SNAPSHOT/org.elasticsearch.action.support.ContextPreservingActionListener.onResponse(ContextPreservingActionListener.java:32)
	at org.elasticsearch.server@8.14.0-SNAPSHOT/org.elasticsearch.action.ActionListenerImplementations$MappedActionListener.onResponse(ActionListenerImplementations.java:95)
	at org.elasticsearch.server@8.14.0-SNAPSHOT/org.elasticsearch.action.ActionListenerResponseHandler.handleResponse(ActionListenerResponseHandler.java:48)
	at org.elasticsearch.server@8.14.0-SNAPSHOT/org.elasticsearch.transport.TransportService$ContextRestoreResponseHandler.handleResponse(TransportService.java:1466)
	at org.elasticsearch.server@8.14.0-SNAPSHOT/org.elasticsearch.transport.TransportService$DirectResponseChannel.processResponse(TransportService.java:1563)
	at org.elasticsearch.server@8.14.0-SNAPSHOT/org.elasticsearch.transport.TransportService$DirectResponseChannel.sendResponse(TransportService.java:1537)
	at org.elasticsearch.server@8.14.0-SNAPSHOT/org.elasticsearch.transport.TaskTransportChannel.sendResponse(TaskTransportChannel.java:35)
	at org.elasticsearch.server@8.14.0-SNAPSHOT/org.elasticsearch.action.support.ChannelActionListener.onResponse(ChannelActionListener.java:32)
	at org.elasticsearch.server@8.14.0-SNAPSHOT/org.elasticsearch.action.support.ChannelActionListener.onResponse(ChannelActionListener.java:19)
	at org.elasticsearch.server@8.14.0-SNAPSHOT/org.elasticsearch.action.ActionListener.respondAndRelease(ActionListener.java:289)
	at org.elasticsearch.server@8.14.0-SNAPSHOT/org.elasticsearch.action.ActionRunnable$3.accept(ActionRunnable.java:78)
	at org.elasticsearch.server@8.14.0-SNAPSHOT/org.elasticsearch.action.ActionRunnable$3.accept(ActionRunnable.java:75)
	at org.elasticsearch.server@8.14.0-SNAPSHOT/org.elasticsearch.action.ActionRunnable$4.doRun(ActionRunnable.java:100)
	at org.elasticsearch.server@8.14.0-SNAPSHOT/org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingAbstractRunnable.doRun(ThreadContext.java:984)
	at org.elasticsearch.server@8.14.0-SNAPSHOT/org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:26)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
	at java.base/java.lang.Thread.run(Thread.java:1570)
"}],
```

</details>

But I can't tell figure out where that call originates in order to do the same in the async-search-status pathway. Any ideas on how to do that are appreciated.